### PR TITLE
Remove calls to agg_and_log_metrics

### PR DESCRIPTION
**Pull request by akashkw**
## What does this PR do?
Removes all calls to agg_and_log_metrics in the codebase. Danielle showed here https://github.com/PyTorchLightning/pytorch-lightning/issues/9145#issuecomment-1026186725 that agg_and_log_metrics is a no-op and can safely be removed.

Part of #9145 

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/loggers/test_base.py [hunk 1]
```diff
@@ -225,31 +225,6 @@ def validation_epoch_end(self, outputs):
     trainer.fit(model)
 
 
-def test_with_accumulate_grad_batches():
```
Comment (akashkw) [inline tests/loggers/test_base.py:228]: Since the aggregation logic is in the training loop, I'm not sure this test makes sense anymore

File: tests/loggers/test_base.py [hunk 2]
```diff
@@ -225,31 +225,6 @@ def validation_epoch_end(self, outputs):
...
-
-    np.random.seed(42)
-    for i, loss in enumerate(np.random.random(10)):
-        logger.agg_and_log_metrics({"loss": loss}, step=int(i / 5))
-
-    assert logger.history == {0: {"loss": 0.5623850983416314}}
-    logger.save()
-    assert logger.history == {0: {"loss": 0.5623850983416314}, 1: {"loss": 0.4778883735637184}}
-
-
```
Comment (akashkw) [inline tests/loggers/test_base.py:252]: Since the aggregation logic is in the training loop, I'm not sure this test makes sense anymore

Comment (daniellepintz) [reply tests/loggers/test_base.py:252]: We could always keep this test but just add a `pytest.deprecated_call` like here: https://github.com/PyTorchLightning/pytorch-lightning/blob/1b107c58928e75b42155aef4a3695b5ee243c276/tests/deprecated_api/test_remove_1-8.py#L46

Comment (akashkw) [reply tests/loggers/test_base.py:252]: Currently the deprecation message is only shown at initialization if someone overrides agg_and_log_metrics in a custom logger. If we put a deprecation message in this method, it would spam it over and over each time it's called. Is that a behavior we're ok with?

Comment (daniellepintz) [reply tests/loggers/test_base.py:252]: Hmm, yeah I see the problem, the way it is now we need a Trainer to emit the dep message. I'm not suggesting we put the dep message in `agg_and_log_metrics`

**Review by daniellepintz — APPROVED** [review]
Looks great to me!

File: pytorch_lightning/utilities/model_helpers.py [hunk 3]
```diff
@@ -52,5 +52,4 @@ def is_overridden(method_name: str, instance: Optional[object] = None, parent: O
     parent_attr = getattr(parent, method_name, None)
     if parent_attr is None:
         raise ValueError("The parent should define the method")
-
```
Comment (ananthsub) [inline pytorch_lightning/utilities/model_helpers.py:55]: Remove

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 4]
```diff
@@ -64,6 +66,14 @@ def on_trainer_init(
         self.trainer.flush_logs_every_n_steps = flush_logs_every_n_steps
         self.trainer.log_every_n_steps = log_every_n_steps
         self.trainer.move_metrics_to_cpu = move_metrics_to_cpu
+        if not isinstance(self.trainer.logger, LoggerCollection) and is_overridden(
+            "agg_and_log_metrics", self.trainer.logger, LightningLoggerBase
+        ):
+            self._override_agg_and_log_metrics = True
+            rank_zero_deprecation(
+                "`LightningLoggerBase.agg_and_log_metrics` is deprecated in v1.6 and will be"
+                " removed in v1.8. Please use `LightningLoggerBase.log_metrics` instead."
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:75]: This should mention that the Trainer will directly call `LightningLoggerBase.log_metrics`

The "user" using the API is the Trainer. The developer may have implemented a logger which implements both log_Metrics & agg_and_log_metrics and would be confused by this

**Review by ananthsub — COMMENTED** [review]
are you planning to deprecate these in a separate PR? 
https://github.com/PyTorchLightning/pytorch-lightning/blob/1515ef90ee2724bcba46e1434eb4b4f9719ebdd7/pytorch_lightning/loggers/base.py#L65-L66
https://github.com/PyTorchLightning/pytorch-lightning/blob/1515ef90ee2724bcba46e1434eb4b4f9719ebdd7/pytorch_lightning/loggers/base.py#L81-L100

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 4 above)
Comment (akashkw) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:75]: Good point, I'll clarify the error message

**Comment by akashkw** [comment]
> are you planning to deprecate these in a separate PR?
> 
> https://github.com/PyTorchLightning/pytorch-lightning/blob/1515ef90ee2724bcba46e1434eb4b4f9719ebdd7/pytorch_lightning/loggers/base.py#L65-L66
> 
> 
> https://github.com/PyTorchLightning/pytorch-lightning/blob/1515ef90ee2724bcba46e1434eb4b4f9719ebdd7/pytorch_lightning/loggers/base.py#L81-L100

Yep, making a separate PR for those.

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 5]
```diff
@@ -64,6 +66,14 @@ def on_trainer_init(
         self.trainer.flush_logs_every_n_steps = flush_logs_every_n_steps
         self.trainer.log_every_n_steps = log_every_n_steps
         self.trainer.move_metrics_to_cpu = move_metrics_to_cpu
+        if not isinstance(self.trainer.logger, LoggerCollection) and is_overridden(
```
[the remark this answers is not in the document]
Comment (akashkw) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:69]: are you talking about the scenario where one of the loggers in trainer.loggers overrides `agg_and_log_metrics`?

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 6]
```diff
@@ -114,7 +125,10 @@ def log_metrics(self, metrics: _OUT_DICT, step: Optional[int] = None) -> None:
             step = self.trainer.global_step
 
         # log actual metrics
-        self.trainer.logger.agg_and_log_metrics(scalar_metrics, step=step)
+        if self._override_agg_and_log_metrics:
+            self.trainer.logger.agg_and_log_metrics(metrics=scalar_metrics, step=step)
```
Comment (daniellepintz) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:129]: for a separate PR: shouldn't we be calling `log_metrics` and `save` on *each* logger here?

Comment (akashkw) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:129]: That's correct. Right now this works because `self.logger` is a `LoggerCollection` but this needs to be changed before `LoggerCollection` is removed.

**Review by ananthsub — APPROVED** [review]

**Comment by akashkw** [comment]
> are you planning to deprecate these in a separate PR?
PR is up now https://github.com/PyTorchLightning/pytorch-lightning/pull/11871

**Comment by akashkw** [comment]
> are you planning to deprecate these in a separate PR?
> 
> https://github.com/PyTorchLightning/pytorch-lightning/blob/1515ef90ee2724bcba46e1434eb4b4f9719ebdd7/pytorch_lightning/loggers/base.py#L65-L66
> 
> 
> https://github.com/PyTorchLightning/pytorch-lightning/blob/1515ef90ee2724bcba46e1434eb4b4f9719ebdd7/pytorch_lightning/loggers/base.py#L81-L100

Yep, PR is up now https://github.com/PyTorchLightning/pytorch-lightning/pull/11871

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/loggers/base.py [hunk 7]
```diff
@@ -126,7 +129,10 @@ def _aggregate_metrics(
         return agg_step, agg_mets
 
     def _reduce_agg_metrics(self):
-        """Aggregate accumulated metrics."""
+        """Aggregate accumulated metrics.
+
+        .. deprecated:: v1.6     This method is deprecated in v1.6 and will be removed in v1.8.
```
Comment (rohitgr7) [inline pytorch_lightning/loggers/base.py:134]:
```suggestion
        .. deprecated:: v1.6
            This method is deprecated in v1.6 and will be removed in v1.8.
```

File: pytorch_lightning/loggers/base.py [hunk 8]
```diff
@@ -137,7 +143,10 @@ def _reduce_agg_metrics(self):
         return self._prev_step, agg_mets
 
     def _finalize_agg_metrics(self):
-        """This shall be called before save/close."""
+        """This shall be called before save/close.
+
+        .. deprecated:: v1.6     This method is deprecated in v1.6 and will be removed in v1.8.
```
Comment (rohitgr7) [inline pytorch_lightning/loggers/base.py:148]:
```suggestion
        .. deprecated:: v1.6
            This method is deprecated in v1.6 and will be removed in v1.8.
```

**Review by rohitgr7 — APPROVED** [review]
nit:
LGTM!

**Review by Borda — DISMISSED** [review]
temp block until @awaelchli checks it, for the context https://github.com/PyTorchLightning/pytorch-lightning/issues/9145#issuecomment-1038498380

File: pytorch_lightning/loggers/base.py (hunk 7 above)
Comment (akashkw) [reply pytorch_lightning/loggers/base.py:134]: Strange... your suggestion is exactly what I wrote but it looks like pre-commit changed it here https://github.com/PyTorchLightning/pytorch-lightning/pull/11832/commits/1b7a1bd05ac2fdecc7c1d765c54e3056b23a28b3

If I commit your suggestion, pre-commit will probably change it back again. I'll play around with the docstring and find something that pre-commit likes.

Comment (carmocca) [reply pytorch_lightning/loggers/base.py:134]: See https://github.com/PyCQA/docformatter/issues/78 for the workaround

File: tests/deprecated_api/test_remove_1-8.py [hunk 9]
```diff
@@ -500,3 +501,41 @@ def on_before_accelerator_backend_setup(self, *args, **kwargs):
         " and will be removed in v1.8"
     ):
         trainer.fit(model)
+
+
+def test_v1_8_0_deprecated_agg_and_log_metrics_override(tmpdir):
+    class AggregationLogger(CSVLogger):
+        def __init__(self, *args, **kwargs):
+            super().__init__(*args, **kwargs)
+
```
Comment (carmocca) [inline tests/deprecated_api/test_remove_1-8.py:510]: Unused init?

```suggestion
```

File: tests/deprecated_api/test_remove_1-8.py [hunk 10]
```diff
@@ -500,3 +501,41 @@ def on_before_accelerator_backend_setup(self, *args, **kwargs):
...
+        @rank_zero_only
+        def agg_and_log_metrics(self, metrics, step):
+            self.log_metrics(metrics=metrics, step=step)
+
+    class NoAggregationLogger(CSVLogger):
+        def __init__(self, *args, **kwargs):
+            super().__init__(*args, **kwargs)
```
Comment (carmocca) [inline tests/deprecated_api/test_remove_1-8.py:517]:
```suggestion
        ...
```

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 11]
```diff
@@ -64,6 +66,15 @@ def on_trainer_init(
         self.trainer.flush_logs_every_n_steps = flush_logs_every_n_steps
         self.trainer.log_every_n_steps = log_every_n_steps
         self.trainer.move_metrics_to_cpu = move_metrics_to_cpu
+        for logger in self.trainer.loggers:
+            if is_overridden("agg_and_log_metrics", logger, LightningLoggerBase):
+                self._override_agg_and_log_metrics = True
+                rank_zero_deprecation(
+                    "`LightningLoggerBase.agg_and_log_metrics` is deprecated in v1.6 and will be removed"
+                    " in v1.8. `Trainer` will directly call `LightningLoggerBase.log_metrics so custom"
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:74]: eye of an eagle!
```suggestion
                    " in v1.8. `Trainer` will directly call `LightningLoggerBase.log_metrics` so custom"
```

**Review by awaelchli — APPROVED** [review]
